index.js

```diff
@@ -18,110 +18,148 @@ function parseDotenvFile(path, verbose = false) {
   return dotenv.parse(content)
 }
 
-module.exports = ({types: t}) => ({
-  name: 'dotenv-import',
-
-  pre() {
-    this.opts = {
-      envName: 'APP_ENV',
-      moduleName: '@env',
-      path: '.env',
-      whitelist: null,
-      blacklist: null,
-      allowlist: null,
-      blocklist: null,
-      safe: false,
-      allowUndefined: true,
-      verbose: false,
-      ...this.opts,
-    }
+module.exports = (api, options) => {
+  const t = api.types
+  this.env = {}
+  options = {
+    envName: 'APP_ENV',
+    moduleName: '@env',
+    path: '.env',
+    whitelist: null,
+    blacklist: null,
+    allowlist: null,
+    blocklist: null,
+    safe: false,
+    allowUndefined: true,
+    verbose: false,
+    ...options,
+  }
+  const babelMode = process.env[options.envName] || (process.env.BABEL_ENV && process.env.BABEL_ENV !== 'undefined' && process.env.BABEL_ENV !== 'development' && process.env.BABEL_ENV) || process.env.NODE_ENV || 'development'
+  const localFilePath = options.path + '.local'
+  const modeFilePath = options.path + '.' + babelMode
+  const modeLocalFilePath = options.path + '.' + babelMode + '.local'
 
-    const babelMode = process.env[this.opts.envName] || (process.env.BABEL_ENV && process.env.BABEL_ENV !== 'undefined' && process.env.BABEL_ENV !== 'development' && process.env.BABEL_ENV) || process.env.NODE_ENV || 'development'
-    if (this.opts.verbose) {
-      console.log('dotenvMode', babelMode)
-    }
+  if (options.verbose) {
+    console.log('dotenvMode', babelMode)
+  }
+
+  api.cache.using(() => {
```

Comment:
The function passed to `api.cache.using` should return a value that, when it changes, causes the plugin to be re-instantiated. Since this one always returns `undefined`, the plugin will never be re-instantiated and it will always use the initial `.env` file.

I suggest doing something like this:
```js
function mtime(filePath) {
  try {
    return fs.statSync(filePath).mtimeMs
  } catch {
    return null
  }
}

api.cache.using(() => mtime(options.path))
api.cache.using(() => mtime(localFilePath))
api.cache.using(() => mtime(modeFilePath))
api.cache.using(() => mtime(modeLocalFilePath))

let env

if (options.safe) {
  const parsed = parseDotenvFile(options.path, options.verbose)
  const localParsed = parseDotenvFile(localFilePath, options.verbose)
  const modeParsed = parseDotenvFile(modeFilePath, options.verbose)
  const modeLocalParsed = parseDotenvFile(modeLocalFilePath, options.verbose)
  
  env = Object.assign(Object.assign(Object.assign(parsed, modeParsed), localParsed), modeLocalParsed)
} else {
  dotenv.config({
    path: modeLocalFilePath,
    silent: true,
  })
  dotenv.config({
    path: modeFilePath,
    silent: true,
  })
  dotenv.config({
    path: localFilePath,
    silent: true,
  })
  dotenv.config({
    path: options.path,
  })
      
  env = process.env
}

```

Reply:
so when it is re-initialized, `pre()` is rerun?

Reply:
so I found that this did not work in tests unfortunately. do I also need to use `addExternalDependency`?